File: exercises/python/lab08/shm_reader.py
"""Lab 08 - Shared Memory IPC (Reader) using mmap + file"""
import struct
import os

SHM_FILE = "/tmp/os_lab08_shm"


def main():
    if not os.path.exists(SHM_FILE):
        print("No shared memory file found. Run shm_writer.py first.")
        return

    with open(SHM_FILE, "rb") as f:
        length = struct.unpack("I", f.read(4))[0]
        data   = f.read(length).decode()

    print(f'Data read from shared memory: "{data}"')
    os.remove(SHM_FILE)
    print("Shared memory file removed.")


if __name__ == "__main__":
    main()

File: exercises/cpp/lab08/shm_reader.cpp
// Lab 08 - Shared Memory IPC (Reader)
#include <iostream>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <cstdio>
using namespace std;

int main() {
    key_t key = ftok("shmfile", 65);
    if (key == -1) { perror("ftok (run shm_writer first)"); return 1; }

    int shmid = shmget(key, 1024, 0666 | IPC_CREAT);
    if (shmid == -1) { perror("shmget"); return 1; }

    char* str = (char*)shmat(shmid, nullptr, 0);
    if (str == (char*)-1) { perror("shmat"); return 1; }

    printf("Data read from shared memory: \"%s\"\n", str);

    shmdt(str);
    // Destroy shared memory segment
    shmctl(shmid, IPC_RMID, nullptr);
    printf("Shared memory segment destroyed.\n");
    return 0;
}
